src/libraries/Common/src/System/IO/FileSystem.DirectoryCreation.Windows.cs

```diff
@@ -26,39 +22,41 @@ public static unsafe void CreateDirectory(string fullPath, byte[]? securityDescr
                 return;
             }
 
-            List<string> stackDir = new List<string>();
+            List<string> stackDir = new();
 
             // Attempt to figure out which directories don't exist, and only
             // create the ones we need.  Note that FileExists may fail due
             // to Win32 ACL's preventing us from seeing a directory, and this
             // isn't threadsafe.
 
-            bool somepathexists = false;
-            int length = fullPath.Length;
+            bool somePathExists = false;
+
+            ReadOnlySpan<char> fullPathSpan = fullPath.AsSpan();
+            int length = fullPathSpan.Length;
 
             // We need to trim the trailing slash or the code will try to create 2 directories of the same name.
-            if (length >= 2 && PathInternal.EndsInDirectorySeparator(fullPath.AsSpan()))
+            if (length >= 2 && PathInternal.EndsInDirectorySeparator(fullPathSpan))
             {
                 length--;
             }
 
-            int lengthRoot = PathInternal.GetRootLength(fullPath.AsSpan());
+            int lengthRoot = PathInternal.GetRootLength(fullPathSpan);
 
             if (length > lengthRoot)
             {
                 // Special case root (fullpath = X:\\)
                 int i = length - 1;
-                while (i >= lengthRoot && !somepathexists)
+                while (i >= lengthRoot && !somePathExists)
                 {
-                    string dir = fullPath.Substring(0, i + 1);
+                    string dir = fullPath[..length];
```

Comment:
Couldn't this be slicing `fullPathSpan` and only allocating a string when actually required to add to `stackDir`?

Reply:
It is already slicing :D, but allocating when its required will be implemented

Reply:
But that code allocates a new string for every slice.

Reply:
That is required, because we call `DirectoryExists` for this https://github.com/dotnet/runtime/pull/62860/files#diff-06fae15f80982c429b14937113da886ffab25f4ef2d3ed39e68df86c5189f7ddR53

The param of `DirectoryExists` must be allocated, because this calls Kernel32 [`FindFileFirst`](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-findfirstfilea)

Reply:
Hmm, I was certain it was taking a `ReadOnlySpan<char>`.

Reply:
Spans can be [passed to native code if they're fixed](https://github.com/dotnet/runtime/issues/27091#issuecomment-411475536)

Not sure whether it's worth wiring through or not here though.